### PR TITLE
Use messages when outputting errors instead of codes

### PR DESCRIPTION
**Pull request by onei**
@TehKittyCat Tested all errors except curl and http errors, filled in some docs as well

### CODE REVIEW

File: RSHighscores.php [hunk 1]
```diff
@@ -20,10 +20,13 @@
...
 	'name'			=> 'RSHiscores',
-	'version'		=> '3.0.1-dev',
+	'version'		=> '3.1.1-dev',
 	'descriptionmsg'	=> 'rshiscores-desc',
 	'url'			=> 'https://github.com/TehKittyCat/RSHiscores',
-	'author'		=> '[http://runescape.wikia.com/wiki/User_talk:TehKittyCat TehKittyCat]',
+	'author'		=> array(
+		'[http://runescape.wikia.com/wiki/User_talk:TehKittyCat TehKittyCat]',
+		'[http://runescape.wikia.com/wiki/User:Cqm Cqm]',
+	),
```
Comment (TehKittyCat) [inline RSHighscores.php]: Probably should add Quarenon as well, he did give design feedback and the initial draft of the caching functionality.

File: RSHighscores.i18n.php [hunk 2]
```diff
@@ -8,10 +8,31 @@
 
 $messages = array();
 
+$messages['qqq'] = array(
+	'rshiscores-desc'		=> '{{desc}}',
+	'rshiscores-error-category'	=> 'Category for tagging errors returned by the parser function.',
+	'rshiscores-error-A'		=> 'Error message for missing player name.',
+	'rshiscores-error-B'		=> 'Error message for a player that was not found in the RuneScape Hiscores',
+	'rshiscores-error-C'		=> 'Error message for when a curl error occurs, whre $1 is the number of the curl error.',
```
Comment (TehKittyCat) [inline RSHighscores.i18n.php]: Typo in where.

**Comment by TehKittyCat** [comment]
I noted a typo and the possibly of adding Quarenon to author for his contributions, but otherwise it looks good. I haven't had a chance to test yet. I assume GitHub notified you, but I think I gave you push access to the repo. Should we migrate RSHiscores to RuneScape Wiki's organization?

**Comment by TehKittyCat** [comment]
We also need to figure how to manage the repositories. Should we maintain branches and when ready do pull requests for the other to review?

File: RSHighscores.php [hunk 3]
```diff
@@ -20,10 +20,13 @@
 $wgExtensionCredits['parserhook'][] = array(
 	'path'			=> __FILE__,
 	'name'			=> 'RSHiscores',
-	'version'		=> '3.0.1-dev',
+	'version'		=> '3.1.1-dev',
```
Comment (TehKittyCat) [inline RSHighscores.php]: I think you meant 3.1.0-dev.

**Comment by onei** [comment]
I made all the changes/fixes you suggested. I did see the push access, but felt it was a good idea to have someone look over it just in case :)

Migrating the repo is apparently done in settings, and it seems like a good idea to have it there if we're using it as a central runescape wiki thng.

I'm largely following the review process wiikia use right now, so I'll push changes to a separate branch ready for reviewing before merging into the master branch. On wikia it's more so broken stuff doesn't end up in production by accident (not really an issue here), but I think it's a sensible process to follow in general.

**Comment by onei** [comment]
Another thought, this can't be submitted to wikia without the memcached fixes as it'll cause the message cache to be rebuilt.

**Comment by onei** [comment]
For testing, I made this as a simple test case:
```
* A: {{#hs:rs3}}
* B: {{#hs:rs3|kjkhdvasn}}
* C: ''skipped''
* D: ''skipped''
* F: {{#hs:rs3|Zezima|100}}
* G: {{#hs:rs3|Zezima|0|5}}
* H: {{#hs:}}
* E: {{#hs:rs3|bar}}
```
Interestingly for H, it doesn't appear that the default value of `'rs3'` ever gets set.

**Comment by TehKittyCat** [comment]
For testing I used:
```
<!-- Functional test (no error code)
{{#hs:rs3|zezima}}
{{#hs:osrs|zezima}} -->

<!-- Cache test (no error code)
{{#hs:rs3|TehKittyCat}}
{{#hs:rs3|TehKittyCat}}
{{#hs:rs3|TehKittyCat}}
-->

<!-- No-name test (A)
{{#hs:rs3|}} -->

<!-- Unlisted player test (B)
{{#hs:rs3|TehKittyNoName}} -->

<!-- Skip C -->
<!-- Skip D -->

<!-- Limit test (E)
{{#hs:rs3|TehKittyCat}}
{{#hs:rs3|zezima}}
{{#hs:rs3|bob}}
-->

<!-- Unknown skill (F)
{{#hs:rs3|Zezima|100}} -->

<!-- Unknown type (G)
{{#hs:rs3|Zezima|0|5}} -->

<!-- Unknown API (H)
{{#hs:rsc}} -->
```

A few notes:
- We should add an error code for when Jagex blocks us as well.
- cURL's error handling is broken.
  - Unexpected curl error returned: .
  - PHP Notice:  Undefined variable: data in RSHighscores.body.php on line 181.
- I didn't test D (wasn't willing to get blocked).
- I didn't find any (other) regressions.
- Otherwise it works great.

I've setup a 1.19.24 test wiki and a trunk test wiki on my system for testing, with memcached, so I'm going to work on that as a separate pull request, then we need to decide on whether to throttle or what else to do. (See my last comment in #3 for an option).

File: RSHighscores.php [hunk 4]
```diff
@@ -20,10 +20,14 @@
 $wgExtensionCredits['parserhook'][] = array(
 	'path'			=> __FILE__,
 	'name'			=> 'RSHiscores',
-	'version'		=> '3.0.1-dev',
+	'version'		=> '3.1.0-dev',
 	'descriptionmsg'	=> 'rshiscores-desc',
 	'url'			=> 'https://github.com/TehKittyCat/RSHiscores',
```
Comment (TehKittyCat) [inline RSHighscores.php]: Just remembered, we need to switch to the new URL.

**Comment by onei** [comment]
A timeout is something I was going to work on when we altered the caching, I was thinking using a new global instead of the default HTTP Timeout, using a default of 5 seconds. For now I'd rather this be as much about error messages as possible before moving onto other things.

Fixed the curl error handling (and HTTP status error handling) and the URL.

**Comment by TehKittyCat** [comment]
Well everything looks good, if you feel it's ready, go ahead and pull it. I switched out the hiscores URL for testing C/D error codes with a simple PHP script returning different error codes. In doing so, I learned the error handling is broken, so that when the status isn't 200 or 404, curl_exec returns false, so I'll work on fixing that separately.

**Pull request merged**